src/v/io/scheduler.cc

```diff
@@ -0,0 +1,162 @@
+    return io_queue_.path();
+}
+
+// NOLINTNEXTLINE(readability-function-cognitive-complexity)
+seastar::future<> scheduler::monitor(queue* queue) noexcept {
+    while (true) {
+        /*
+         * cooperative scheduling for cache eviction. consider the case when
+         * nofiles_ semaphore is initialized with 0 units (weird, but ok, maybe
+         * we want to be able to pause all I/O). two queues block on nofiles_
+         * semaphore. later a unit is deposited into the semaphore and the first
+         * wakes up and opens its underlying io_queue. at this point the running
+         * queue monitor could be suspended, but we'd like to be able for all
+         * queues to make progress. to handle this the active queue monitor will
+         * evict one or more other queues before suspending itself. however,
+         * because an io queue needs to wake up to be closed there is a delay
+         * before the eviction is visible. this delay is why we track waiters_
+         * separately rather than using the nofiles_ semaphore state.
+         */
+        while (waiters_ > 0 && !lru_.empty()) {
+            waiters_--;
+            auto& queue = lru_.front();
+            lru_.pop_front();
+            queue.sem_.signal();
+        }
+
+        /*
+         * when stopping the queue clean it up fully including removing it from
+         * the cache and closing the underlying io-queue.
+         */
+        if (queue->stop_) {
+            if (queue->io_queue_.opened()) {
+                if (queue->cache_hook_.is_linked()) {
+                    queue->cache_hook_.unlink();
+                }
+                co_await queue->io_queue_.close();
+                queue->units_.return_all();
+            }
+            co_return;
+        }
+
+        co_await queue->sem_.wait(std::max<size_t>(queue->sem_.current(), 1));
+
+        /*
+         * in the common case the queue is already open so there is nothing else
+         * to do and the monitor can loop around to suspend itself. however, if
+         * the queue has also been removed from the lru cache list then this is
+         * a signal that the opened queue should be closed and nofiles_ units
+         * released.
+         */
+        if (queue->io_queue_.opened()) {
+            if (!queue->cache_hook_.is_linked()) {
+                co_await queue->io_queue_.close();
+                queue->units_.return_all();
+            }
+            continue;
+        }
+
+        /*
+         * open the queue. if we can't get nofiles units then we'll try to
+         * evict another queue and wait for units to become available.
+         */
+        auto units = seastar::try_get_units(nofiles_, 1);
+        if (!units.has_value()) {
+            if (lru_.empty()) {
+                /*
+                 * this case is possible, if for example, `nofiles_` was
+                 * initialized with 0 units.
+                 */
+                waiters_++;
```

Comment:
I'm a bit confused by this: I suspect a subsequent change will allow changing the semaphore to be adjustable dynamically, so for now, it's odd to see. Assuming adjusting the semaphore will be the way we'd be able to unblock the `get_units` call, shouldn't we also be decrementing `waiters_` once we add ourselves to `lru_`?

Reply:
i added a comment about waiters being used to handle cases that arise when the semaphore units are dynamically adjusted.

>shouldn't we also be decrementing waiters_ once we add ourselves to lru_?

I think that the net effect is that this is happening, but we can't easily have a worker decrement uncondtionally when it gets added to the lru list (even if it knows it contributed a +1) because its hard to know if another worker woekup and woke up some waiters (the cooperative wake up at the top is a while loop that wakes up as much as possible). not being careful would have us double count decrements.

i'm not convinced its 100% correct, but here is a simple example:

say we have open file limit as 0 and three queues show up. lru is empty, so waiters will bump to 3. then we deposit one unit into the semaphpore and a waiter wakes up. it'll add itself to the lru and loop back around and wake up waiters. the waiters are still 3, but the lru only has 1 entry, so so that limits the evictions to 1 (correctly, the currently running worker is the only open file and now itll self-close so another worker runs). waiters is decremented to 2 reflecting reality.

Reply:
Ah thanks for the example! Very helpful. I'd skipped over the fact that we can still proceed with opening without `waiters_` being decremented, as long as `open_file_limit_` is added to. I initially thought another monitor had to pop from the `lru_` for us to proceed, which was confusing because `lru_` here is empty!

Reply:
right. i guess waiters has kind of a optimistic property in that it's incremented before the waiter waits, and decremented before the waiter releases.